internal/image/digest_selector.go

```diff
@@ -36,52 +38,33 @@ func newDigestSelector(
 
 // Select implements the Selector interface.
 func (d *digestSelector) Select(ctx context.Context) ([]Image, error) {
+	tag := d.constraint
 	logger := logging.LoggerFromContext(ctx).WithFields(log.Fields{
 		"registry":            d.repoClient.registry.name,
-		"image":               d.repoClient.image,
+		"image":               d.repoClient.repoURL,
 		"selectionStrategy":   SelectionStrategyDigest,
+		"tag":                 tag,
 		"platformConstrained": d.platform != nil,
 	})
 	logger.Trace("selecting image")
 
 	ctx = logging.ContextWithLogger(ctx, logger)
 
-	// TODO(hidde): it would be much more efficient to directly attempt
-	// to retrieve the image for the tag, while gracefully handling the
-	// case where it does not exist. This would avoid the need to list
-	// all tags and then iterate over them.
-	tags, err := d.repoClient.getTags(ctx)
+	image, err := d.repoClient.getImageByTag(ctx, tag, d.platform)
 	if err != nil {
-		return nil, fmt.Errorf("error listing tags: %w", err)
-	}
-	if len(tags) == 0 {
-		logger.Trace("found no tags")
-		return nil, nil
-	}
-	logger.Trace("got all tags")
-
-	for _, tag := range tags {
-		if tag != d.constraint {
-			continue
-		}
-		image, err := d.repoClient.getImageByTag(ctx, tag, d.platform)
-		if err != nil {
-			return nil, fmt.Errorf("error retrieving image with tag %q: %w", tag, err)
-		}
-		if image == nil {
-			logger.Tracef(
-				"image with tag %q was found, but did not match platform constraint",
-				tag,
-			)
+		var te *transport.Error
+		if errors.As(err, &te) && te.StatusCode == http.StatusNotFound {
+			logger.Trace("found no image with tag")
 			return nil, nil
 		}
-		logger.WithFields(log.Fields{
-			"tag":    image.Tag,
-			"digest": image.Digest.String(),
-		}).Trace("found image")
-		return []Image{*image}, nil
+		return nil, fmt.Errorf("error retrieving image with tag %q: %w", tag, err)
+	}
+
+	if image == nil {
+		logger.Trace("image with tag did not match platform constraints", tag)
```

Comment:
```suggestion
		logger.Trace("image with tag did not match platform constraints")
```

Reply:
Thanks... I was using `logger.Tracef()` for a hot minute and then added `tag` to the structured part of the log instead. Looks like I botched it when setting this back to `logger.Trace()`. I'm a bit surprised no linter caught this.

Reply:
Linter typically only knows about the std libraries accepting format strings, which means it will just think this is a normal string here without looking further at what's in it.